### PR TITLE
MGT - CI

### PR DESCRIPTION
**Pull request by kshxtij**
Implement CI 

- [x] Build on Mac and Ubuntu
- [x] Ctest
- [x] Fix logging test to work (github servers are in a different timezone 😀)
- [ ] Code Coverage
- [x] Clang-Format
- [x] Todo & Fixme check (TODOLater for preexisting todos that are harder to fix)

### CODE REVIEW

**Comment by kshxtij** [comment]
Could fix the timezone error using this snippet within the tests

```c++
  setenv("TZ", "Europe/London", 1);
  tzset();
```

but this isn't the most elegant solution

**Review by mifrandir — APPROVED** [review]
Looks good except for the timezone stuff. What we need to find out is how to refer to 1970-01-01 00:00 in the local time zone; this would give us an offset in seconds which we could use in the tests. `std::chrono::parse` is supposed to solve this but I can't make it work; I believe it's not properly implemented yet.

**Comment by kshxtij** [comment]
I think a stopgap for now could be just setting the timezone within the test and revisiting this

**Comment by mifrandir** [comment]
> Thoughts on how to get clang-fomat as a CI action. What i'm considering rn is, run on pullrequest and comment everything that's wrong.
> 
> I'm open to suggestions though, @mifrandir @TomLonergan03 @maxguy2001 @SnickeyX
> 
> Edit: Could also combine clang-tidy with clang-format into one action

You can dry run clang-format with the -Werror flag. This happens in scripts/hooks/pre-commit. This will fail if things aren't formatted properly.

**Comment by kshxtij** [comment]
Example of on-pull-request action running clang-format and cpptidy, https://github.com/kshxtij/hyped-2023/pull/2#issuecomment-1297205710

**Comment by mifrandir** [comment]
> Example of on-pull-request action running clang-format and cpptidy, [kshxtij#2 (comment)](https://github.com/kshxtij/hyped-2023/pull/2#issuecomment-1297205710)

oh that's what you mean by comment; thats quite cool actually

**Comment by mifrandir** [comment]
Oops, this was meant to go on a different branch. Ah well...

**Review by mifrandir — APPROVED** [review]

**Comment by kshxtij** [comment]
@maxguy2001 @SnickeyX @arjunnaha could i get quick reviews on this, want this merged before the next set of PRs are being worked on.

I know code coverage isn't added yet, but i'll come back and fix that over the weekend.

**Review by SnickeyX — APPROVED** [review]
Looks good

**Review by maxguy2001 — APPROVED** [review]

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -0,0 +1,27 @@
+name: Build
+
+on: push
```
Comment (arjunnaha) [inline .github/workflows/build.yml:3]: Do you want this to be _every_ push? I think `pull_request` to master is better. And then you can make `master` a protected branch requiring a PR and CI to succeed

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+        with:
+          compiler: clang
+          ubuntu: |
+            apt: libeigen3-dev
+          macos: |
+            brew: eigen
+      - name: CMake
+        run: cmake -S . -B build -DCMAKE_CXX_COMPILER=$(which clang++)
+      - name: Build hyped
+        run: cd build; make -j
```
Comment (arjunnaha) [inline .github/workflows/build.yml:25]: [`working_directory`](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idstepsrun) might be slightly more syntactic than `cd` but this is just very nitpicky ;)

**Review by TomLonergan03 — APPROVED** [review]

**Review by arjunnaha — APPROVED** [review]
LGTM

**Pull request merged**